src/bin/adbackup-cli.rs

```diff
@@ -19,41 +21,137 @@ fn main() {
+            .long("device")
+            .short("d")
+            .takes_value(true)
+            .value_name("ID")
+    };
+
     App::new("adbackup")
         .about("A backup tool for android using adb")
         .author(crate_authors!())
         .version(adbackup::version())
         .setting(AppSettings::SubcommandRequiredElseHelp)
-        .arg(Arg::with_name("verbose")
-            .short("v")
-            .multiple(true)
-            .global(true)
-            .help("Increases logging verbosity each use for up to 3 times"))
-        .subcommand(SubCommand::with_name("devices")
-            .about("List connected devices")
-            .help("List all android devices connected to your pc with enabled debug mode."))
+        .arg(
+            Arg::with_name("verbose")
+                .short("v")
+                .multiple(true)
+                .global(true)
+                .help("Increases logging verbosity each use for up to 3 times"),
+        )
+        .subcommand(
+            SubCommand::with_name("backup")
+                .display_order(1)
+                .about("Start backup of device")
+                .arg(device_arg())
+                .arg(
+                    Arg::with_name("applications")
+                        .help("Include the apk's into the backup")
+                        .long("applications")
+                        .short("applications"),
+                )
+                .arg(
+                    Arg::with_name("shared")
+                        .help("Include the shared storage into the backup")
+                        .long("shared")
+                        .short("s"),
+                )
+                .arg(
+                    Arg::with_name("system")
+                        .help("Include the system apps storage into the backup")
+                        .long("system")
+                        .short("S"),
+                )
+                .arg(
+                    Arg::with_name("only_specified")
+                        .help("Include only the specified apps into the backup")
+                        .long("specified")
+                        .short("o")
+                        .takes_value(true)
+                        .multiple(true)
+                        .value_name("APP"),
+                ),
+        )
+        .subcommand(
+            SubCommand::with_name("devices")
+                .display_order(2)
+                .about("List connected devices")
+                .arg(device_arg())
+                .help("List all android devices connected to your pc with enabled debug mode."),
+        )
+        .subcommand(
+            SubCommand::with_name("apps")
+                .display_order(3)
+                .about("List all installed apps on devices")
+                .arg(device_arg()),
+        )
 }
 
-fn print_devices() -> Result<(), Error> {
+fn print_devices(_: &ArgMatches, _: Option<&ArgMatches>) -> Result<(), Error> {
```

Comment:
why are these params needed?

Reply:
It is required because if we return from the match a fuction and call one of them, all functions must have the same parameters:
``` rust
    let sub_fn = match sub_name {
        "devices" => print_devices,
        "backup" => backup,
        "apps" => apps,
        _ => unimplemented!(),
    }
    ...
    let result = sub_fn(&matches, subm);
```

I think there are not that much functions, which do not require parameters because of that I thought it would be the best solution but yes it is not perfect, do you have a better one?

Reply:
maybe call the functions in the match directly? we could then make the params more specific for each of them (not sure if your new functions always need all matches and subm?!). even if we are calling most of them with the same args, I would find this approach a little cleaner than giving functions unnecessary params